src/error.rs

```diff
@@ -4,83 +4,265 @@ use std::io;
+    BadResponse(String),
+    /// The server returned an error prefixed with SERVER_ERROR in response to a command.
+    Error(String),
+}
+
+impl fmt::Display for ServerError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ServerError::BadMagic(e) => write!(f, "Expected 0x81 as magic in response header, but found: {:x}", e),
+            ServerError::BadResponse(s) => write!(f, "Unexpected: {} in response", s),
+            ServerError::Error(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+/// Command specific errors.
+#[derive(Debug, PartialEq)]
+pub enum CommandError {
+    /// The client tried to set a key which already existed in the server.
+    KeyExists,
+    /// The client tried to set a key which does not exist in the server.
+    KeyNotFound,
+    /// The value for a key was too large. The limit is usually 1MB.
+    ValueTooLarge,
+    /// Invalid arguments were passed to the command.
+    InvalidArguments,
+    /// The server requires authentication.
+    AuthenticationRequired,
+    /// When using binary protocol, the server returned an unknown response status.
+    Unknown(u16),
+    /// The client sent an invalid command to the server.
+    InvalidCommand,
+}
+
+impl MemcacheError {
+    pub(crate) fn try_from(s: String) -> Result<String, MemcacheError> {
+        if s == "ERROR\r\n" {
+            Err(CommandError::InvalidCommand)?
+        } else if s.starts_with("CLIENT_ERROR") {
+            Err(ClientError::from(s))?
+        } else if s.starts_with("SERVER_ERROR") {
+            Err(ServerError::from(s))?
+        } else if s == "NOT_FOUND\r\n" {
+            Err(CommandError::KeyNotFound)?
+        } else if s == "EXISTS\r\n" {
+            Err(CommandError::KeyExists)?
+        } else {
+            Ok(s)
+        }
+    }
+}
+
+impl From<String> for ClientError {
+    fn from(s: String) -> Self {
+        ClientError::Error(s)
+    }
+}
+
+impl From<String> for ServerError {
+    fn from(s: String) -> Self {
+        ServerError::Error(s)
+    }
+}
+
+impl fmt::Display for CommandError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            CommandError::KeyExists => write!(f, "Key already exists in the server."),
+            CommandError::KeyNotFound => write!(f, "Key was not found in the server."),
+            CommandError::ValueTooLarge => write!(f, "Value was too large."),
+            CommandError::InvalidArguments => write!(f, "Invalid arguments provided."),
+            CommandError::AuthenticationRequired => write!(f, "Authentication required."),
+            CommandError::Unknown(code) => write!(f, "Unknown error occurred with code: {}.", code),
+            CommandError::InvalidCommand => write!(f, "Invalid command sent to the server."),
+        }
+    }
+}
+
+impl From<u16> for CommandError {
+    fn from(status: u16) -> CommandError {
+        match status {
+            0x0 => panic!("shouldn't reach here"),
```

Comment:
I think we should avoid using `panic!` for do not exit user's process no matter what memcached responsed (maybe some one is using this crate for custom service which talks memcache protocol).

Reply:
This is done mostly for convenience. `0x0` status is success case and we will very likely catch any panics during development phase. I'll change the panic message, though.

Reply:
For now I think it's ok, but I think in future refactors, someone may breaking this by accident. Although this should be a bug, and user got a unexcpeted error, but I think it's better to crash user's process. I think we can just return a ClientError with this message.

Reply:
Changed the panics to errors.